contribute/design/1.x/ndm/20200219-ndm-disk-accessing.md

```diff
@@ -0,0 +1,184 @@
+  on `/dev/sdb` and it has Major and Minor number of disk as `8:16` then detached
+  disk1 and attached disk2 to the same node disk2 access path came as same `/dev/sdb`
+  and Maj:Min number will be `8:16` which will be not recommended to use them.
+  So to tackle this created loop device on `disk1` access path and given to the
+  application(application starts consuming disk via block file). Now the disk1 was
+  detached and attached back to the node kernel will allocate new Major & Minor
+  number and new access path `/dev/sdx(next available path)` to the same disk because
+  loop device is holding old disk path & Maj:Min number] in the tempfs(Ex: /tmp) + user specified path +
+  uniquely identified id[
+  Why input from the user? Because multiple application can consume blockdevice so
+  it is recommended to provide some unique path to block file.
+  Why tempfs? Upon reboots of the nodes, loop devices will be destroyed because
+  of devfs is temporary kind of filesystem up on reboots.
+  Why uniquely identified id? The application can identify the disk using that
+  unique id]. This entire path will be updated on blockdevice status(New field).
+  So the user can access the disk with the path that exists on the status of blockdevice.
+
+- Now if an application requires multiple blockdevices application should make a
+  the claim of all the blockdevices with some common path on blockdeviceclaim.
+
+##### Workflow
+
+- NDM will create blockdevices for uniquely identified blockdevices when NDM pod
+  up or when the disk was attached.
+- The application will makes a request to claim the blockdevice using blockdeviceclaim.
+  User will mention path on blockdeviceclaim(new field) on which path it needs to
+  access the blockdevice.
+- NDM will create a loop device and block file on the requested blockdevice.
+  Ex: To create loop device command used -- `losetup loop0 /dev/disk/by-id/path`
+      To create block file command used -- `mknod <path_to_file> b Maj Min`
+- Once the above step is successful NDM will update the status as `Claimed` and
+  path to access block devices.
+- When the existing claimed blockdevice was detached and attached back during
+  updating the status as `Active` NDM should create new loop device on new path
+  recreate block file to point it to new Major and Minor number(Application
+  consuming blockdevice should reopen the ports because file got recreated on a
+  same path with the same name).
+- Upon reboots of the node NDM should create blockfiles and loop devices for
+  claimed disks in the specified path.
+
+##### Advantages
+
+- Application can uniquely access the disks.
+- Application doesn't required to run in privileged mode to access the disks.
+- Dynamically attached disks can be easily consumed by the application without
+  restarting application pod.
+- Application consuming multiple blockdevices no need to mount all of them (or)
+  no need to mount the /dev directory inside the application.
```

Comment:
@mittachaitu . Suppose the application creates a partition on the disk. How will the application access partition it created ?

Reply:
It will be possible with the following approach but not sure how far it is correct:
1. Given block file(pointing to entier disk) to the application by mounting blockfile directory(which contains the only disk it can access) into the container.
2. Application created partitions on that disk.
3. The application has to claim a partitioned block device(by specifying the same access path whatever specified to claim entier disk).
4. Now the path to access the partition exists inside the container. 
Not sure whether it is correct or not just kept my thoughts.

Reply:
Currently in Kubernetes world accessing partitioned disk that was created by the application itself is not straight forward.